examples/DDIM/model.jl

```diff
@@ -0,0 +1,364 @@
+# DDIM implementation follwoing https://keras.io/examples/generative/ddim/
+
+using Lux
+using Random
+using CUDA
+
+# Note: Julia/Lux assume image batch of WHCN ordering
+
+# Embed noise variances to embedding
+function sinusoidal_embedding(x::AbstractArray{T, 4}, min_freq::T, max_freq::T,
+                              embedding_dims::Int) where {T <: AbstractFloat}
+    if size(x)[1:3] != (1, 1, 1)
+        throw(DimensionMismatch("Input shape must be (1, 1, 1, batch)"))
+    end
+
+    # define frequencies
+    # LinRange requires @adjoint when used with Zygote
+    # Instead we manually implement range.
+    lower = log(min_freq)
+    upper = log(max_freq)
+    n = div(embedding_dims, 2)
+    d = (upper - lower) / (n - 1)
+    freqs = exp.(lower:d:upper) |> gpu
+    @assert length(freqs) == div(embedding_dims, 2)
+    @assert size(freqs) == (div(embedding_dims, 2),)
+
+    angular_speeds = reshape(convert(T, 2) * π * freqs, (1, 1, length(freqs), 1))
+    @assert size(angular_speeds) == (1, 1, div(embedding_dims, 2), 1)
+
+    embeddings = cat(sin.(angular_speeds .* x), cos.(angular_speeds .* x); dims=3)
+    @assert size(embeddings) == (1, 1, embedding_dims, size(x, 4))
+
+    return embeddings
+end
+
+# activations
+function sigmoid(x)
+    return 1 / (1 + exp(-x))
+end
+
+function silu(x)
+    return x * sigmoid(x)
+end
+
+# Basic building block of UNet
+function residual_block(in_channels::Int, out_channels::Int)
+    if in_channels == out_channels
+        first_layer = NoOpLayer()
+    else
+        first_layer = Conv((3, 3), in_channels => out_channels; pad=SamePad())
+    end
+
+    return Lux.Chain(first_layer,
+                     Lux.SkipConnection(Lux.Chain(BatchNorm(out_channels; affine=false,
+                                                            momentum=0.99),
+                                                  Conv((3, 3), out_channels => out_channels;
+                                                       stride=1, pad=(1, 1)),
+                                                  WrappedFunction(Base.Fix1(broadcast,
+                                                                            silu)),
+                                                  Conv((3, 3), out_channels => out_channels;
+                                                       stride=1, pad=(1, 1))), +))
+end
+
+# Downsampling block of UNet
+# It narrows height and width while increasing channels.
+struct DownBlock <: Lux.AbstractExplicitContainerLayer{(:residual_blocks, :maxpool)}
+    residual_blocks::Lux.Chain
```

Comment:
Too tight a type requirement. If it is a `AbstractExplicitLayer` it should work.

Reply:
I modified `Chain` to `AbstractExplicitLayer`